File: pylox/pylox/Lox.py
# /bin/env python3
import sys
from typing import overload

from Token import Token
from TokenTypes import TokenType
from RuntimeError import RuntimeError
from Interpreter import Interpreter
from Parser import Parser
from Scanner import Scanner
from Stmt import Stmt
from LoxCallable import LoxCallable
from LoxFunction import LoxFunction
from Resolver import Resolver
from LoxClass import LoxClass


class Lox:
    _interpreter: Interpreter = Interpreter(LoxCallable, LoxFunction, LoxClass)
    _had_error = False
    _had_runtime_error = False

    @staticmethod
    def _run_file(path: str):
        with open(path, encoding="utf-8") as f:
            Lox._run(f.read())
        if Lox._had_error:
            sys.exit(65)
        if Lox._had_runtime_error:
            sys.exit(70)

    @staticmethod
    def _run_prompt():
        while True:
            line = input(">")
            if not line:
                break
            Lox._run(line)
            Lox._had_error = False

    @staticmethod
    def _run(source: str):
        scanner: Scanner = Scanner(source, Lox.error)
        tokens: list[Token] = scanner.scan_tokens()
        parser: Parser = Parser(tokens, Lox.error)
        statements: list[Stmt] = parser.parse()

        if Lox._had_error:
            return

        resolver: Resolver = Resolver(Lox._interpreter, Lox.error)
        resolver.resolve(statements)

        if Lox._had_error:
            return

        Lox._interpreter.interpret(statements, Lox.runtime_error)

    @overload
    @staticmethod
    def error(line: int, message: str) -> None: ...

    @overload
    @staticmethod
    def error(token: Token, message: str) -> None: ...

    @staticmethod
    def error(arg: int | Token, message: str) -> None:  # type: ignore[reportInconsistentOverload]
        if isinstance(arg, int):
            Lox._report(arg, "", message)
        else:
            if arg.type == TokenType.EOF:
                Lox._report(arg.line, " at end", message)
            else:
                Lox._report(arg.line, f" at '{arg.lexeme}'", message)

    @staticmethod
    def runtime_error(error: RuntimeError):
        print(f"{error}\n[line {error.token.line}]", file=sys.stderr)
        Lox._had_runtime_error = True

    @staticmethod
    def _report(line: int, where: str, message: str):
        print(f"[line {line}] Error{where}: {message}", file=sys.stderr)
        Lox._had_error = True

    @staticmethod
    def main() -> None:
        if len(sys.argv) > 2:
            print("Usage: pylox [script]", file=sys.stderr)
            sys.exit(64)
        elif len(sys.argv) == 2:
            Lox._run_file(sys.argv[1])
        else:
            Lox._run_prompt()


if __name__ == "__main__":
    Lox.main()

File: golox/lox.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"

	"github.com/englishlayup/crafting_interpreters/golox/internal/interpreter"
	"github.com/englishlayup/crafting_interpreters/golox/internal/parser"
	"github.com/englishlayup/crafting_interpreters/golox/internal/scanner"
	"github.com/englishlayup/crafting_interpreters/golox/internal/token"
)

type Interpreter = interpreter.Interpreter

type Lox struct {
	interpreter     Interpreter
	hadError        bool
	hadRuntimeError bool
}

func (lox *Lox) runFile(path string) {
	fileData, err := os.ReadFile(path)
	if err != nil {
		log.Fatal(err)
	}

	lox.run(string(fileData))

	if lox.hadError {
		os.Exit(65)
	}
	if lox.hadRuntimeError {
		os.Exit(70)
	}
}

func (lox *Lox) runPrompt() {
	reader := bufio.NewReader(os.Stdin)
	for {
		fmt.Print(">")
		input, err := reader.ReadString('\n')
		if err != nil {
			log.Fatal(err)
		}
		lox.run(input)
		lox.hadError = false
	}
}

func (lox *Lox) run(source string) error {
	scanner := scanner.NewScanner(source, lox.reportScannerError)
	tokens := scanner.ScanTokens()
	parser := parser.NewParser(tokens, lox.reportError)
	statements := parser.Parse()
	for _, t := range statements {
		fmt.Print(t)
	}

	fmt.Print(nil)
	return nil
}

func (lox *Lox) reportScannerError(line int, message string) {
	lox.report(line, "", message)
}

func (lox *Lox) reportError(t token.Token, message string) {
	if t.TokenType == token.EOF {
		lox.report(t.Line, " at end", message)
	} else {
		lox.report(t.Line, fmt.Sprintf(" at '%v'", t.Lexeme), message)
	}
}

func (lox *Lox) report(line int, where string, message string) {
	fmt.Fprintf(os.Stderr, "[line %v] Error%v: %v", line, where, message)
	lox.hadError = true
}
